typescript/schedules.md

```diff
@@ -0,0 +1,1255 @@
+  | 0
+  | 1
+  | 2
+  | 3
+  | 4
+  | 5
+  | 6
+  | 7
+  | 8
+  | 9
+  | 10
+  | 11
+  | 12
+  | 13
+  | 14
+  | 15
+  | 16
+  | 17
+  | 18
+  | 19
+  | 20
+  | 21
+  | 22
+  | 23
+  | 24
+  | 25
+  | 26
+  | 27
+  | 28
+  | 29
+  | 30
+  | 31
+  | 32
+  | 33
+  | 34
+  | 35
+  | 36
+  | 37
+  | 38
+  | 39
+  | 40
+  | 41
+  | 42
+  | 43
+  | 44
+  | 45
+  | 46
+  | 47
+  | 48
+  | 49
+  | 50
+  | 51
+  | 52
+  | 53
+  | 54
+  | 55
+  | 56
+  | 57
+  | 58
+  | 59;
+
+export type Second = Range<ZeroTo59> | ZeroTo59;
+export type Minute = Range<ZeroTo59> | ZeroTo59;
+
+export type ZeroTo23 =
+  | 0
+  | 1
+  | 2
+  | 3
+  | 4
+  | 5
+  | 6
+  | 7
+  | 8
+  | 9
+  | 10
+  | 11
+  | 12
+  | 13
+  | 14
+  | 15
+  | 16
+  | 17
+  | 18
+  | 19
+  | 20
+  | 21
+  | 22
+  | 23;
+
+export type Hour = Range<ZeroTo23> | ZeroTo23;
+
+export type OneTo31 =
+  | 1
+  | 2
+  | 3
+  | 4
+  | 5
+  | 6
+  | 7
+  | 8
+  | 9
+  | 10
+  | 11
+  | 12
+  | 13
+  | 14
+  | 15
+  | 16
+  | 17
+  | 18
+  | 19
+  | 20
+  | 21
+  | 22
+  | 23
+  | 24
+  | 25
+  | 26
+  | 27
+  | 28
+  | 29
+  | 30
+  | 31;
+
+export type DayOfMonth = Range<OneTo31> | OneTo31;
+
+export type OneTo12 = 1 | 2 | 3 | 4 | 5 | 6 | 7 | 8 | 9 | 10 | 11 | 12;
+
+export type MonthName =
+  | 'JAN'
+  | 'JANUARY'
+  | 'FEB'
+  | 'FEBRUARY'
+  | 'MAR'
+  | 'MARCH'
+  | 'APR'
+  | 'APRIL'
+  | 'MAY'
+  | 'JUN'
+  | 'JUNE'
+  | 'JUL'
+  | 'JULY'
+  | 'AUG'
+  | 'AUGUST'
+  | 'SEP'
+  | 'SEPTEMBER'
+  | 'OCT'
+  | 'OCTOBER'
+  | 'NOV'
+  | 'NOVEMBER'
+  | 'DEC'
+  | 'DECEMBER';
+
+export type MonthValue = OneTo12 | MonthName;
+
+export type Month = Range<MonthValue> | MonthValue;
+
+export type TwoThousandTo2100 =
+  | 2000
+  | 2001
+  | 2002
+  | 2003
+  | 2004
+  | 2005
+  | 2006
+  | 2007
+  | 2008
+  | 2009
+  | 2010
+  | 2011
+  | 2012
+  | 2013
+  | 2014
+  | 2015
+  | 2016
+  | 2017
+  | 2018
+  | 2019
+  | 2020
+  | 2021
+  | 2022
+  | 2023
+  | 2024
+  | 2025
+  | 2026
+  | 2027
+  | 2028
+  | 2029
+  | 2030
+  | 2031
+  | 2032
+  | 2033
+  | 2034
+  | 2035
+  | 2036
+  | 2037
+  | 2038
+  | 2039
+  | 2040
+  | 2041
+  | 2042
+  | 2043
+  | 2044
+  | 2045
+  | 2046
+  | 2047
+  | 2048
+  | 2049
+  | 2050
+  | 2051
+  | 2052
+  | 2053
+  | 2054
+  | 2055
+  | 2056
+  | 2057
+  | 2058
+  | 2059
+  | 2060
+  | 2061
+  | 2062
+  | 2063
+  | 2064
+  | 2065
+  | 2066
+  | 2067
+  | 2068
+  | 2069
+  | 2070
+  | 2071
+  | 2072
+  | 2073
+  | 2074
+  | 2075
+  | 2076
+  | 2077
+  | 2078
+  | 2079
+  | 2080
+  | 2081
+  | 2082
+  | 2083
+  | 2084
+  | 2085
+  | 2086
+  | 2087
+  | 2088
+  | 2089
+  | 2090
+  | 2091
+  | 2092
+  | 2093
+  | 2094
+  | 2095
+  | 2096
+  | 2097
+  | 2098
+  | 2099
+  | 2100;
+
+/**
+ * Temporal Server currently only supports specifying years in the range `[2000, 2100]`.
+ */
+export type Year = Range<TwoThousandTo2100> | TwoThousandTo2100;
+
+/**
+ * Both 0 and 7 mean Sunday
+ */
+export type ZeroTo7 = 0 | 1 | 2 | 3 | 4 | 5 | 7;
+
+export type DayOfWeekName =
```

Comment:
Same as month, these should be an enum IMO, not a fixed set of literals. And we shouldn't support ambiguity or abbreviations, we should require accurate values.

Reply:
Why not be as flexible as the API? (which is equalsIgnoreCase 2+ characters)

Reply:
> Why not be as flexible as the API?

Just saw they accept any string which I think is unfortunate. But therefore if you choose to be as flexible as the API, you should also accept any string. But what you have here is _intentionally_ not as flexible as the API type. I think if you're gonna be stricter at typecheck time here, you might as well really be stricter.

Reply:
+1 to requiring accurate values

Reply:
the api works with strings because in many situations (tctl, web ui) users will read and write these schedule specs directly in json format, and we didn't want to make that overly verbose, or require an sdk just to be able to write "friday".

I'm all in favor of the sdks enforcing more structure and precision on things.

Reply:
Service interprets 0 and 7 as Sunday (noted in comment). We could be US-centric and decide that weeks start on Sunday and only support 0-6?

Reply:
For the string inputs like this, which of these options do we like?

1. Have single value (enum or string) per day, and dev writes logic to convert end user input into our value
2. Type it `string` and either SDK or server throws at runtime if wrong

2 seems like less work for dev?

Reply:
I vote for string union for TS / JS.
No need to accept numbers, it's just confusing.

Reply:
I'm looking at it like: if 2 or more of our users implement "sun" or number parsing for their end users, then it's more efficient if we do it once, for them?

Reply:
Which we could do by either accepting `string` or giving them a `parseDayOfWeek` helper that converts "sun" to our "SUNDAY"

Reply:
Meeting notes:

month and dayOfWeek string types:

server accepts: `{  dayOfWeek: 'sun' or '2' }`

use for now: `dayOfWeek: string`

server: change month and dayOfWeek to only accept canonical? don't have time to do before 1.18 release. keep it beta so API can change. david will think on it. talk to ryland.

then we can do `dayOfWeek: 'SUNDAY' | 'MONDAY'`

server doesn't support:
- `* * sun * 2`

sdk could support?
- "extended cron syntax" `every 2 weeks`. does current cron support this? go sdk supports some.
  - current alternative: `{interval: { every: '2wk'}}`? more likely to use calendar.
  
maybe new schedule description offset from current server time. "every 2 weeks" starting now. can add. 

can also add X seconds after previous finishes. 👍